src/Avalonia.Base/Input/TextInput/ITextInputMethodClient.cs

```diff
@@ -10,57 +8,81 @@ public interface ITextInputMethodClient
         /// </summary>
         event EventHandler? CursorRectangleChanged;
+
         /// <summary>
         /// The visual that's showing the text
         /// </summary>
         Visual TextViewVisual { get; }
+
         /// <summary>
         /// Should be fired when text-hosting visual is changed
         /// </summary>
         event EventHandler? TextViewVisualChanged;
+
         /// <summary>
         /// Indicates if TextViewVisual is capable of displaying non-committed input on the cursor position
         /// </summary>
         bool SupportsPreedit { get; }
+
         /// <summary>
         /// Sets the non-committed input string
         /// </summary>
         void SetPreeditText(string? text);
 
-        /// <summary>
-        /// Sets the current composing region. This doesn't remove the composing text from the commited text.
-        /// </summary>
-        void SetComposingRegion(TextRange? region);
-
         /// <summary>
         /// Indicates if text input client is capable of providing the text around the cursor
         /// </summary>
         bool SupportsSurroundingText { get; }
+
         /// <summary>
         /// Returns the text around the cursor, usually the current paragraph, the cursor position inside that text and selection start position
         /// </summary>
-        TextInputMethodSurroundingText SurroundingText { get; }
+        string? SurroundingText { get; }
+
         /// <summary>
         /// Should be fired when surrounding text changed
         /// </summary>
         event EventHandler? SurroundingTextChanged;
 
         /// <summary>
-        /// Gets or sets a platform editable. Text and selection changes made in the editable are forwarded to the IM client.
+        /// Gets or sets he selection range within current surrounding text.
         /// </summary>
-        ITextEditable? TextEditable { get; set; }
+        TextSelection Selection { get; set; }
 
-        void SelectInSurroundingText(int start, int end);
-    }
+        /// <summary>
+        /// Should be fired when the selection changed
+        /// </summary>
+        event EventHandler? SelectionChanged;
 
-    public record struct TextInputMethodSurroundingText
-    {
-        public string Text { get; set; }
-        public int CursorOffset { get; set; }
-        public int AnchorOffset { get; set; }
+        /// <summary>
+        /// Inserts text at current cursor position.
+        /// </summary>
+        /// <param name="text">The inserted text</param>
+        /// <remarks>
+        /// Replaces currently selected text range.
+        /// </remarks>
+        void Insert(string? text);
+
+        /// <summary>
+        /// Gets or sets the composition region.
+        /// </summary>
+        /// <remarks>
+        /// This is used to highlight currently active composition region.
+        /// </remarks>
+        CompositionRegion CompositionRegion { get; set; }
```

Comment:
Shouldn't the composition be controlled by the input method?

Reply:
This state is shared by the platform IME and the TextPresenter. The region should only be set by the IME. This is mainly used to highlight the currently active composition region. We can split this into `CompositionRegion { get; }` `SetCompositionRegion(CompositionRegion)` Like PreeditText

Reply:
I have changed it to just expose SetCompositionRegion(CompositionRegion)

Reply:
Removed